File: internal/dash/api/stats.go
package dash_api

import (
	"fmt"
	"net/http"
	"time"

	"github.com/MunifTanjim/stremthru/internal/anilist"
	"github.com/MunifTanjim/stremthru/internal/cache"
	"github.com/MunifTanjim/stremthru/internal/config"
	"github.com/MunifTanjim/stremthru/internal/db"
	"github.com/MunifTanjim/stremthru/internal/imdb_title"
	"github.com/MunifTanjim/stremthru/internal/letterboxd"
	"github.com/MunifTanjim/stremthru/internal/magnet_cache"
	"github.com/MunifTanjim/stremthru/internal/mdblist"
	"github.com/MunifTanjim/stremthru/internal/serializd"
	"github.com/MunifTanjim/stremthru/internal/shared"
	"github.com/MunifTanjim/stremthru/internal/tmdb"
	"github.com/MunifTanjim/stremthru/internal/torrent_info"
	"github.com/MunifTanjim/stremthru/internal/torrent_stream"
	torznab_indexer_syncinfo "github.com/MunifTanjim/stremthru/internal/torznab/indexer/syncinfo"
	"github.com/MunifTanjim/stremthru/internal/trakt"
	"github.com/MunifTanjim/stremthru/internal/tvdb"
	store_stats "github.com/MunifTanjim/stremthru/store/stats"
)

var cachedTorrentsStats = cache.NewCachedValue(cache.CachedValueConfig[*torrent_info.Stats]{
	Get: torrent_info.GetStats,
	TTL: 6 * time.Hour,
})

type CacheStatsEntry struct {
	Hit  int64 `json:"hit"`
	Miss int64 `json:"miss"`
}

type TorrentsStats struct {
	TotalCount int `json:"total_count"`
	Files      struct {
		TotalCount int `json:"total_count"`
	} `json:"files"`
	Cache struct {
		WriteTorrentInfo   CacheStatsEntry `json:"write_torrent_info"`
		ReadTorrentStream  CacheStatsEntry `json:"read_torrent_stream"`
		WriteTorrentStream CacheStatsEntry `json:"write_torrent_stream"`
		ReadMagnetCache    CacheStatsEntry `json:"read_magnet_cache"`
		WriteMagnetCache   CacheStatsEntry `json:"write_magnet_cache"`
	} `json:"cache"`
}

func HandleGetTorrentsStats(w http.ResponseWriter, r *http.Request) {
	if !shared.IsMethod(r, http.MethodGet) {
		ErrorMethodNotAllowed(r).Send(w, r)
		return
	}

	stats, err := cachedTorrentsStats.Get()
	if err != nil {
		SendError(w, r, err)
		return
	}

	data := TorrentsStats{}
	data.TotalCount = stats.TotalCount
	data.Files.TotalCount = stats.Streams.TotalCount

	tiWriteHit, tiWriteMiss := torrent_info.GetUpsertCacheStats()
	tsReadHit, tsReadMiss := torrent_stream.GetReadCacheStats()
	tsWriteHit, tsWriteMiss := torrent_stream.GetWriteCacheStats()
	mcReadHit, mcReadMiss := magnet_cache.GetReadCacheStats()
	mcWriteHit, mcWriteMiss := magnet_cache.GetWriteCacheStats()
	data.Cache.WriteTorrentInfo = CacheStatsEntry{Hit: tiWriteHit, Miss: tiWriteMiss}
	data.Cache.ReadTorrentStream = CacheStatsEntry{Hit: tsReadHit, Miss: tsReadMiss}
	data.Cache.WriteTorrentStream = CacheStatsEntry{Hit: tsWriteHit, Miss: tsWriteMiss}
	data.Cache.ReadMagnetCache = CacheStatsEntry{Hit: mcReadHit, Miss: mcReadMiss}
	data.Cache.WriteMagnetCache = CacheStatsEntry{Hit: mcWriteHit, Miss: mcWriteMiss}

	SendData(w, r, 200, data)
}

type IMDBTitleStats struct {
	TotalCount int `json:"total_count"`
}

var cachedIMDBTitleStats = cache.NewCachedValue(cache.CachedValueConfig[*IMDBTitleStats]{
	Get: func() (*IMDBTitleStats, error) {
		var count int
		err := db.QueryRow(fmt.Sprintf(`SELECT COUNT(1) FROM %s`, imdb_title.TableName)).Scan(&count)
		if err != nil {
			return nil, err
		}
		return &IMDBTitleStats{
			TotalCount: count,
		}, nil
	},
	TTL: 6 * time.Hour,
})

func HandleGetIMDBTitleStats(w http.ResponseWriter, r *http.Request) {
	if !shared.IsMethod(r, http.MethodGet) {
		ErrorMethodNotAllowed(r).Send(w, r)
		return
	}

	stats, err := cachedIMDBTitleStats.Get()
	if err != nil {
		SendError(w, r, err)
		return
	}

	SendData(w, r, 200, stats)
}

type ServerStatsFeature struct {
	IMDBTitle bool `json:"imdb_title"`
	Meta      bool `json:"meta"`
	Newz      bool `json:"newz"`
	Torz      bool `json:"torz"`
	Sync      bool `json:"sync"`
	Vault     bool `json:"vault"`
}

type ServerStatsIntegration struct {
	Trakt bool `json:"trakt"`
}

type ServerStats struct {
	Version     string                 `json:"version"`
	StartedAt   time.Time              `json:"started_at"`
	Integration ServerStatsIntegration `json:"integration"`
}

func HandleGetServerStats(w http.ResponseWriter, r *http.Request) {
	if !shared.IsMethod(r, http.MethodGet) {
		ErrorMethodNotAllowed(r).Send(w, r)
		return
	}

	data := ServerStats{
		Version:   config.Version,
		StartedAt: config.ServerStartTime,
		Integration: ServerStatsIntegration{
			Trakt: config.Integration.Trakt.IsEnabled(),
		},
	}
	SendData(w, r, 200, data)
}

type ListsStats struct {
	AniList struct {
		TotalLists int `json:"total_lists"`
		TotalItems int `json:"total_items"`
	} `json:"anilist"`
	Letterboxd struct {
		TotalLists int `json:"total_lists"`
		TotalItems int `json:"total_items"`
	} `json:"letterboxd"`
	MDBList struct {
		TotalLists int `json:"total_lists"`
		TotalItems int `json:"total_items"`
	} `json:"mdblist"`
	TMDB struct {
		TotalLists int `json:"total_lists"`
		TotalItems int `json:"total_items"`
	} `json:"tmdb"`
	Trakt struct {
		TotalLists int `json:"total_lists"`
		TotalItems int `json:"total_items"`
	} `json:"trakt"`
	TVDB struct {
		TotalLists int `json:"total_lists"`
		TotalItems int `json:"total_items"`
	} `json:"tvdb"`
	Serializd struct {
		TotalLists int `json:"total_lists"`
		TotalItems int `json:"total_items"`
	} `json:"serializd"`
}

var query_get_lists_stats = fmt.Sprintf(`
SELECT COUNT(1) FROM %s UNION ALL SELECT COUNT(1) FROM %s
UNION ALL
SELECT COUNT(1) FROM %s UNION ALL SELECT COUNT(1) FROM %s
UNION ALL
SELECT COUNT(1) FROM %s UNION ALL SELECT COUNT(1) FROM %s
UNION ALL
SELECT COUNT(1) FROM %s UNION ALL SELECT COUNT(1) FROM %s
UNION ALL
SELECT COUNT(1) FROM %s UNION ALL SELECT COUNT(1) FROM %s
UNION ALL
SELECT COUNT(1) FROM %s UNION ALL SELECT COUNT(1) FROM %s
UNION ALL
SELECT COUNT(1) FROM %s UNION ALL SELECT COUNT(1) FROM %s
`,
	anilist.ListTableName, anilist.MediaTableName,
	letterboxd.ListTableName, letterboxd.ItemTableName,
	mdblist.ListTableName, mdblist.ItemTableName,
	tmdb.ListTableName, tmdb.ItemTableName,
	trakt.ListTableName, trakt.ItemTableName,
	tvdb.ListTableName, tvdb.ItemTableName,
	serializd.ListTableName, serializd.ItemTableName,
)

var cachedListsStats = cache.NewCachedValue(cache.CachedValueConfig[*ListsStats]{
	Get: func() (*ListsStats, error) {
		rows, err := db.Query(query_get_lists_stats)
		if err != nil {
			return nil, err
		}
		defer rows.Close()

		counts := make([]int, 0, 14)
		for rows.Next() {
			var count int
			if err := rows.Scan(&count); err != nil {
				return nil, err
			}
			counts = append(counts, count)
		}

		stats := ListsStats{}
		stats.AniList.TotalLists = counts[0]
		stats.AniList.TotalItems = counts[1]
		stats.Letterboxd.TotalLists = counts[2]
		stats.Letterboxd.TotalItems = counts[3]
		stats.MDBList.TotalLists = counts[4]
		stats.MDBList.TotalItems = counts[5]
		stats.TMDB.TotalLists = counts[6]
		stats.TMDB.TotalItems = counts[7]
		stats.Trakt.TotalLists = counts[8]
		stats.Trakt.TotalItems = counts[9]
		stats.TVDB.TotalLists = counts[10]
		stats.TVDB.TotalItems = counts[11]
		stats.Serializd.TotalLists = counts[12]
		stats.Serializd.TotalItems = counts[13]

		return &stats, nil
	},
	TTL: 3 * time.Hour,
})

func HandleGetListsStats(w http.ResponseWriter, r *http.Request) {
	if !shared.IsMethod(r, http.MethodGet) {
		ErrorMethodNotAllowed(r).Send(w, r)
		return
	}

	stats, err := cachedListsStats.Get()
	if err != nil {
		SendError(w, r, err)
		return
	}

	SendData(w, r, 200, stats)
}

type TorznabIndexerStats struct {
	TotalCount  int64                                       `json:"total_count"`
	SyncedCount int64                                       `json:"synced_count"`
	QueuedCount int64                                       `json:"queued_count"`
	ErrorCount  int64                                       `json:"error_count"`
	ResultCount int64                                       `json:"result_count"`
	Indexers    []torznab_indexer_syncinfo.IndexerSyncStats `json:"indexers"`
}

var cachedTorznabIndexerStats = cache.NewCachedValue(cache.CachedValueConfig[*TorznabIndexerStats]{
	Get: func() (*TorznabIndexerStats, error) {
		indexers, err := torznab_indexer_syncinfo.GetStats()
		if err != nil {
			return nil, err
		}

		stats := &TorznabIndexerStats{
			Indexers: indexers,
		}
		for _, idx := range indexers {
			stats.TotalCount += idx.TotalCount
			stats.SyncedCount += idx.SyncedCount
			stats.QueuedCount += idx.QueuedCount
			stats.ErrorCount += idx.ErrorCount
			stats.ResultCount += idx.ResultCount
		}

		return stats, nil
	},
	TTL: 1 * time.Hour,
})

func HandleGetTorznabIndexerStats(w http.ResponseWriter, r *http.Request) {
	if !shared.IsMethod(r, http.MethodGet) {
		ErrorMethodNotAllowed(r).Send(w, r)
		return
	}

	stats, err := cachedTorznabIndexerStats.Get()
	if err != nil {
		SendError(w, r, err)
		return
	}

	SendData(w, r, 200, stats)
}

type StoreStatsResponse struct {
	Stores []store_stats.StoreStatsSnapshot `json:"stores"`
}

func HandleGetStoreStats(w http.ResponseWriter, r *http.Request) {
	if !shared.IsMethod(r, http.MethodGet) {
		ErrorMethodNotAllowed(r).Send(w, r)
		return
	}

	window := 1 * time.Hour
	stores := store_stats.GetSnapshot(window)
	if stores == nil {
		stores = []store_stats.StoreStatsSnapshot{}
	}

	SendData(w, r, 200, StoreStatsResponse{Stores: stores})
}

File: apps/dash/src/api/stats.ts
import { useQuery } from "@tanstack/react-query";

import { api } from "@/lib/api";

type IMDBTitleStats = {
  total_count: number;
};

type ListsStats = Record<
  | "anilist"
  | "letterboxd"
  | "mdblist"
  | "serializd"
  | "tmdb"
  | "trakt"
  | "tvdb",
  {
    total_items: number;
    total_lists: number;
  }
>;

type ServerStats = {
  integration: {
    trakt: boolean;
  };
  started_at: string;
  version: string;
};

type TorrentsStats = {
  cache: {
    read_magnet_cache: { hit: number; miss: number };
    read_torrent_stream: { hit: number; miss: number };
    write_magnet_cache: { hit: number; miss: number };
    write_torrent_info: { hit: number; miss: number };
    write_torrent_stream: { hit: number; miss: number };
  };
  files: {
    total_count: number;
  };
  total_count: number;
};

type TorznabIndexerStats = {
  error_count: number;
  indexers: TorznabIndexerSyncStats[];
  queued_count: number;
  result_count: number;
  synced_count: number;
  total_count: number;
};

type TorznabIndexerSyncStats = {
  error_count: number;
  indexer_id: number;
  indexer_name: string;
  last_synced_at: null | string;
  queued_count: number;
  result_count: number;
  synced_count: number;
  total_count: number;
};

const HOUR = 60 * 60 * 1000;

export type StoreStatsEntry = {
  methods: StoreMethodStats[];
  name: string;
};

type StoreMethodStats = {
  avg_duration_ms: number;
  error_count: number;
  error_rate: number;
  max_duration_ms: number;
  min_duration_ms: number;
  name: string;
  p50_duration_ms: number;
  p95_duration_ms: number;
  p99_duration_ms: number;
  requests_per_minute: number;
  total_count: number;
};

type StoreStats = {
  stores: StoreStatsEntry[];
};

export function useIMDBTitleStats() {
  return useQuery({
    queryFn: async () => {
      const { data } = await api<IMDBTitleStats>("/stats/imdb-titles");
      return data;
    },
    queryKey: ["/stats/imdb-titles"],
    staleTime: 2 * HOUR,
  });
}

export function useListsStats() {
  return useQuery({
    queryFn: getListsStats,
    queryKey: ["/stats/lists"],
    staleTime: 2 * HOUR,
  });
}

export function useServerStats() {
  return useQuery({
    queryFn: getServerStats,
    queryKey: ["/stats/server"],
    staleTime: 2 * HOUR,
  });
}

export function useTorrentsStats() {
  return useQuery({
    queryFn: getTorrentsStats,
    queryKey: ["/stats/torrents"],
    retry: false,
    staleTime: 2 * HOUR,
  });
}

export function useTorznabIndexerStats() {
  return useQuery({
    queryFn: async () => {
      const { data } = await api<TorznabIndexerStats>(
        "/stats/torznab-indexers",
      );
      return data;
    },
    queryKey: ["/stats/torznab-indexers"],
    staleTime: 1 * HOUR,
  });
}

const MINUTE = 60 * 1000;

export function useStoreStats() {
  return useQuery({
    queryFn: async () => {
      const { data } = await api<StoreStats>("/stats/stores");
      return data;
    },
    queryKey: ["/stats/stores"],
    refetchInterval: 1 * MINUTE,
    staleTime: 30 * 1000,
  });
}

async function getListsStats() {
  const { data } = await api<ListsStats>("/stats/lists");
  return data;
}

async function getServerStats() {
  const { data } = await api<ServerStats>("/stats/server");
  return data;
}

async function getTorrentsStats() {
  const { data } = await api<TorrentsStats>("/stats/torrents");
  return data;
}
